### PR TITLE
Support XPU ABI=0 build

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130110

# Motivation
This PR intends to support ABI=0 build for XPU backend.

# Additional Context
The major change is adding a compilation option `-D__INTEL_PREVIEW_BREAKING_CHANGES` for the host compiler(gcc) and `-fpreview-breaking-changes` for XPU device kernel code compiler(icpx), why?
Because we use
- gcc to compile host code and link SYCL runtime. So we need to pass `-D__INTEL_PREVIEW_BREAKING_CHANGES` to tell the host compiler invoking the ABI-neutral API included in SYCL. And
- use icpx to compile device kernel code and link SYCL runtime. So we need to pass `-fpreview-breaking-changes` to tell the device kernel compiler building ABI-neutral code. Besides,
- `libsycl-preview.so` is an ABI-neutral library but `libsycl.so` is not.

This PR depends on https://github.com/pytorch/pytorch/pull/131643.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -90,6 +90,9 @@ if(USE_XPU)
     message(WARNING "Not compiling with XPU. Could NOT find SYCL."
     "Suppress this warning with -DUSE_XPU=OFF.")
     caffe2_update_option(USE_XPU OFF)
+  else()
+    # Workaround about XPU ABI neutral build
+    string(APPEND CMAKE_CXX_FLAGS " -D__INTEL_PREVIEW_BREAKING_CHANGES")
```
Comment (EikanWang) [inline cmake/Dependencies.cmake:95]: Please get commitment from the compiler team.

Comment (dvrogozh) [reply cmake/Dependencies.cmake:95]: Can you, please, also get more details into PR description? what "ABI neutral build" means? what is sycl-preview library? will ever changes from preview available in stock?

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -90,6 +90,11 @@ if(USE_XPU)
     message(WARNING "Not compiling with XPU. Could NOT find SYCL."
     "Suppress this warning with -DUSE_XPU=OFF.")
     caffe2_update_option(USE_XPU OFF)
+  else()
+    if(LINUX)
```
Comment (min-jean-cho) [inline cmake/Dependencies.cmake:94]: Why Linux only? Let's do the same for Windows.

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Review by dvrogozh — CHANGES_REQUESTED** [review]
Folks, please, update PR description. I honestly don't understand what ABI=0 means. Change also introduces a number of build flags such as `__INTEL_PREVIEW_BREAKING_CHANGES`, `-fpreview-breaking-changes` and changes link library from `sycl` to `sycl-preview`. These changes looks dangerous and go unexplained in the PR description.

Will there be using any features from sycl stack which might be changed in the backward incompatible way in the future? if yes: what exactly do we plan to use? and why?

**Comment by guangyey** [comment]
> Folks, please, update PR description. I honestly don't understand what ABI=0 means. Change also introduces a number of build flags such as `__INTEL_PREVIEW_BREAKING_CHANGES`, `-fpreview-breaking-changes` and changes link library from `sycl` to `sycl-preview`. These changes looks dangerous and go unexplained in the PR description.
> 
> Will there be using any features from sycl stack which might be changed in the backward incompatible way in the future? if yes: what exactly do we plan to use? and why?

Updated.

**Review by dvrogozh — CHANGES_REQUESTED** [review]
Here is a link to dpcpp side documentation for `-fpreview-breaking-changes`: https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-1/fpreview-breaking-changes.html
Quoting: "This option lets a user tell the compiler that they are willing to give up backward compatibility guarantees and lets the compiler enable new backward breaking changes that will appear in the next major release."

With above, I do not understand why you are making these changes. I see you updated PR description, but it does not give any clarity. You basically just stated that you introduced `-fpreview-breaking-changes` as a major change. Given explanation "because we use gcc to compile host code and link SYCL runtime" is not clear. That's what you are using already as far as I know and everything works fine. If everything already works - why the change?

The flag you introduce seems dangerous for release production builds and must be explained, discussed and agreed by maintainers. It clearly enables some new features in the dpcpp compiler which will be breaking changes in next version. **Which "preview" features you are actually using or going to use?** Unfortunately I do not see any code change in this PR which gives an answer to this question.

**Comment by EikanWang** [comment]
> Here is a link to dpcpp side documentation for `-fpreview-breaking-changes`: https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-1/fpreview-breaking-changes.html Quoting: "This option lets a user tell the compiler that they are willing to give up backward compatibility guarantees and lets the compiler enable new backward breaking changes that will appear in the next major release."
> 
> With above, I do not understand why you are making these changes. I see you updated PR description, but it does not give any clarity. You basically just stated that you introduced `-fpreview-breaking-changes` as a major change. Given explanation "because we use gcc to compile host code and link SYCL runtime" is not clear. That's what you are using already as far as I know and everything works fine. If everything already works - why the change?
> 
> The flag you introduce seems dangerous for release production builds and must be explained, discussed and agreed by maintainers. It clearly enables some new features in the dpcpp compiler which will be breaking changes in next version. **Which "preview" features you are actually using or going to use?** Unfortunately I do not see any code change in this PR which gives an answer to this question.

Guangye has clearly explained why we need to introduce this change and elaborate on the changes. We need to support ABI=0 mode. However, the Intel GPU toolchain only supported ABI=1 by default in the past. In terms of ABI=0 support, the toolchain has to introduce some non-backward-compatible changes from the toolchain product perspective. But it does not matter PyTorch. Please check with the compiler team to get more detailed information.

**Comment by EikanWang** [comment]
@atalman , @albanD , @malfet , may I know your comments on this PR?

**Comment by dvrogozh** [comment]
@EikanWang : When you are referring to support XPU ABI=0, do you actually refer to `_GLIBCXX_USE_CXX11_ABI=0|1`? With the explanation given at https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html?

**Comment by EikanWang** [comment]
> @EikanWang : When you are referring to support XPU ABI=0, do you actually refer to `_GLIBCXX_USE_CXX11_ABI=0|1`? With the explanation given at https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html?

Yes. This PRs intends to support ABI both 0 and 1.

File: cmake/Modules/FindMKLDNN.cmake [hunk 3]
```diff
@@ -36,18 +36,22 @@ IF(NOT MKLDNN_FOUND)
...
     ProcessorCount(proc_cnt)
-    if ((DEFINED ENV{MAX_JOBS}) AND ("$ENV{MAX_JOBS}" LESS_EQUAL ${proc_cnt}))
+    if((DEFINED ENV{MAX_JOBS}) AND ("$ENV{MAX_JOBS}" LESS_EQUAL ${proc_cnt}))
       list(APPEND DNNL_MAKE_COMMAND "-j" "$ENV{MAX_JOBS}")
       if(CMAKE_GENERATOR MATCHES "Make|Ninja")
         list(APPEND DNNL_MAKE_COMMAND "--" "-l" "$ENV{MAX_JOBS}")
       endif()
     endif()
+    if(LINUX)
+      set(ABI_NEUTRAL_FLAGS -fpreview-breaking-changes)
```
Comment (dvrogozh) [inline cmake/Modules/FindMKLDNN.cmake:46]: Do we know the version of dpcpp compiler where required preview changes will land? if we know it, then I think we should have version check here to use or not to use this flag. This is in assumption that we chose not to compile with "preview" changes if we can avoid that.

File: cmake/Modules/FindSYCLToolkit.cmake [hunk 4]
```diff
@@ -51,7 +51,7 @@ find_file(
 # Find SYCL library fullname.
 find_library(
   SYCL_LIBRARY
-  NAMES sycl
+  NAMES sycl-preview
```
Comment (dvrogozh) [inline cmake/Modules/FindSYCLToolkit.cmake:54]: Same as above: if we know dpcpp version where required changes will land, suggest to add version check.

**Review by albanD — COMMENTED** [review]
Do we expect end users to build XPU from source at this moment? Or it is expected they have to use binaries we provide?

**Comment by EikanWang** [comment]
> Do we expect end users to build XPU from source at this moment? Or it is expected they have to use binaries we provide?

@albanD , we intend to support both scenarios and do not limit users' behaviors.

By the way, I think PT 2.5 may still release binaries that require **ABI=0**(https://github.com/pytorch/pytorch/issues/123649). However, PyTorch XPU only supports **ABI=1** at this moment w/o this PR.

**Comment by albanD** [comment]
> we intend to support both scenarios and do not limit users' behaviors.

If we expect users to build from source, then I would expect some documentation (I guess in CONTRIBUTING.md) that explains the environment you expect them to have and the quirks related to this change (as they would need some preview version of intel's lib?).

**Comment by EikanWang** [comment]
> > we intend to support both scenarios and do not limit users' behaviors.
> 
> If we expect users to build from source, then I would expect some documentation (I guess in CONTRIBUTING.md) that explains the environment you expect them to have and the quirks related to this change (as they would need some preview version of intel's lib?).

Hi @albanD , The `libsycl-preview` library is part of the Intel GPU software package, so users do not need to download or install an additional package. It supports both `_GLIBCXX_USE_CXX11_ABI=0` and `_GLIBCXX_USE_CXX11_ABI=1`, whereas `libsycl` only supports `ABI=1`. However, it introduces some non-backward-compatible changes to support `_GLIBCXX_USE_CXX11_ABI=0`. That's why the library is packaged with `-preview` suffix — to inform users about the new changes and provide time for adoption and provide the compatibility for deployment.

**Review by albanD — APPROVED** [review]
Sounds good!

**Comment by guangyey** [comment]
@albanD Thanks very much!

**Pull request closed without merging**